### PR TITLE
[MC-1406] Sophos Central - Get Endpoints in Group action

### PR DESCRIPTION
**Pull request by dsliwinski-r7**
### Description

Describe the proposed changes:

  - Add Get Endpoints in Group action

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "",
    "status": "ok",
    "meta": {},
    "output": {
      "items": [
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "tenant": {
            "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0"
          },
          "hostname": "DESKTOP-CGGBP7V",
          "health": {
            "overall": "bad",
            "threats": {
              "status": "good"
            },
            "services": {
              "status": "bad",
              "serviceDetails": [
                {
                  "name": "HitmanPro.Alert service",
                  "status": "running"
                },
                {
                  "name": "Sophos Anti-Virus",
                  "status": "missing"
                },
                {
                  "name": "Sophos AutoUpdate Service",
                  "status": "running"
                },
                {
                  "name": "Sophos Clean Service",
                  "status": "running"
                },
                {
                  "name": "Sophos Endpoint Defense",
                  "status": "running"
                },
                {
                  "name": "Sophos Endpoint Defense Service",
                  "status": "running"
                },
                {
                  "name": "Sophos File Scanner",
                  "status": "running"
                },
                {
                  "name": "Sophos File Scanner Service",
                  "status": "running"
                },
                {
                  "name": "Sophos IPS",
                  "status": "running"
                },
                {
                  "name": "Sophos MCS Agent",
                  "status": "running"
                },
                {
                  "name": "Sophos MCS Client",
                  "status": "running"
                },
                {
                  "name": "Sophos Network Threat Protection",
                  "status": "missing"
                },
                {
                  "name": "Sophos Safestore Service",
                  "status": "running"
                },
                {
                  "name": "Sophos System Protection Service",
                  "status": "running"
                },
                {
                  "name": "Sophos Web Intelligence Filter Service",
                  "status": "missing"
                },
                {
                  "name": "Sophos Web Intelligence Service",
                  "status": "missing"
                }
              ]
            }
          },
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          },
          "ipv4Addresses": [
            "198.51.100.1"
          ],
          "macAddresses": [
            "01:02:23:24:75:B6"
          ],
          "group": {
            "name": "Example Group",
            "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0"
          },
          "associatedPerson": {
            "name": "DESKTOP-CGGBP7V\\Peppa Pig",
            "viaLogin": "DESKTOP-CGGBP7V\\Peppa Pig",
            "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0"
          },
          "tamperProtectionEnabled": true,
          "lastSeenAt": "2023-01-03T07:45:34.802Z",
          "encryption": {
            "volumes": [
              {
                "volumeId": "\\\\?\\Volume{0037ff91-0000-0000-0000-300300000000}\\",
                "status": "notSupported"
              }
            ]
          },
          "isolation": {
            "status": "notIsolated",
            "adminIsolated": false,
            "selfIsolated": false
          }
        },
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "tenant": {
            "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0"
          },
          "hostname": "DESKTOP-9ISGF63",
          "health": {
            "overall": "bad",
            "threats": {
              "status": "good"
            },
            "services": {
              "status": "good",
              "serviceDetails": [
                {
                  "name": "HitmanPro.Alert service",
                  "status": "running"
                },
                {
                  "name": "Sophos Anti-Virus",
                  "status": "running"
                },
                {
                  "name": "Sophos AutoUpdate Service",
                  "status": "running"
                },
                {
                  "name": "Sophos Clean Service",
                  "status": "running"
                },
                {
                  "name": "Sophos Endpoint Defense",
                  "status": "running"
                },
                {
                  "name": "Sophos Endpoint Defense Service",
                  "status": "running"
                },
                {
                  "name": "Sophos File Scanner",
                  "status": "running"
                },
                {
                  "name": "Sophos File Scanner Service",
                  "status": "running"
                },
                {
                  "name": "Sophos MCS Agent",
                  "status": "running"
                },
                {
                  "name": "Sophos MCS Client",
                  "status": "running"
                },
                {
                  "name": "Sophos Network Threat Protection",
                  "status": "running"
                },
                {
                  "name": "Sophos Safestore Service",
                  "status": "running"
                },
                {
                  "name": "Sophos System Protection Service",
                  "status": "running"
                },
                {
                  "name": "Sophos Web Intelligence Filter Service",
                  "status": "running"
                },
                {
                  "name": "Sophos Web Intelligence Service",
                  "status": "running"
                }
              ]
            }
          },
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          },
          "ipv4Addresses": [
            "198.51.100.1"
          ],
          "macAddresses": [
            "01:02:23:24:75:B6"
          ],
          "group": {
            "name": "Example Group",
            "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0"
          },
          "associatedPerson": {
            "name": "Morpheus",
            "viaLogin": "DESKTOP-9ISGF63\\Anakin Skywalker",
            "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0"
          },
          "tamperProtectionEnabled": true,
          "lastSeenAt": "2022-09-29T08:42:07.536Z",
          "encryption": {
            "volumes": [
              {
                "volumeId": "\\\\?\\Volume{a49cb493-0000-0000-0000-300300000000}\\",
                "status": "notSupported"
              }
            ]
          },
          "isolation": {
            "status": "isolated",
            "adminIsolated": true,
            "selfIsolated": false
          }
        },
```

<summary>
docker run --rm -i rapid7/sophos_central:4.4.0 run < tests/get_endpoints_in_group.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "",
    "status": "ok",
    "meta": {},
    "output": {
      "items": [
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "hostname": "DESKTOP-9ISGF63",
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          }
        },
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "hostname": "DESKTOP-CGGBP7V",
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          }
        }
      ],
      "pages": {
        "size": 50,
        "maxSize": 500
      }
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/sophos_central:4.4.0 run < tests/get_endpoints_in_group_selected_fields.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "",
    "status": "ok",
    "meta": {},
    "output": {
      "items": [
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "hostname": "DESKTOP-CGGBP7V",
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          }
        }
      ],
      "pages": {
        "size": 50,
        "maxSize": 500
      }
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/sophos_central:4.4.0 run < tests/get_endpoints_in_group_search_by_hostname.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "",
    "status": "ok",
    "meta": {},
    "output": {
      "items": [
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "hostname": "DESKTOP-CGGBP7V",
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          }
        },
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "hostname": "DESKTOP-9ISGF63",
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          }
        }
      ],
      "pages": {
        "size": 50,
        "total": 1,
        "items": 2,
        "maxSize": 500
      }
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/sophos_central:4.4.0 run < tests/get_endpoints_in_group_search_by_os_name.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "",
    "status": "ok",
    "meta": {},
    "output": {
      "items": [
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "hostname": "DESKTOP-CGGBP7V",
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          }
        }
      ],
      "pages": {
        "nextKey": "WyIxMDhmMzc4YS01MTE2LTQ3ODktYTNiYi1jZDFjMDc5NjY5ZmYiXQ==",
        "size": 1,
        "total": 2,
        "items": 2,
        "maxSize": 500
      }
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/sophos_central:4.4.0 run < tests/get_endpoints_in_group_first_page.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "",
    "status": "ok",
    "meta": {},
    "output": {
      "items": [
        {
          "id": "9de5069c-5afe-602b-2ea0-a04b66beb2c0",
          "type": "computer",
          "hostname": "DESKTOP-9ISGF63",
          "os": {
            "isServer": false,
            "platform": "windows",
            "name": "Windows 10 Home N",
            "majorVersion": 10,
            "minorVersion": 0,
            "build": 19042
          }
        }
      ],
      "pages": {
        "fromKey": "WyIxMDhmMzc4YS01MTE2LTQ3ODktYTNiYi1jZDFjMDc5NjY5ZmYiXQ==",
        "size": 1,
        "total": 2,
        "items": 2,
        "maxSize": 500
      }
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/sophos_central:4.4.0 run < tests/get_endpoints_in_group_second_page.json
</summary>
</details>

### CODE REVIEW

File: plugins/sophos_central/help.md [hunk 1]
```diff
@@ -1582,7 +1692,7 @@ _This plugin does not contain any troubleshooting information._
 
 # Version History
 
-* 4.4.0 - Add new actions: `Isolate Endpoint`, `Get Endpoint Groups`, `Add Endpoint Group`, `Get Endpoint Group`, `Add Endpoint to Group`, `Remove Endpoint from Group`, `Get Allowed Items`, `Add Allowed Item`, `Remove Allowed Item`, `Get Blocked Items`, `Add Blocked Item`, `Remove Blocked Item`
+* 4.4.0 - Add new actions: `Isolate Endpoint`, `Get Endpoint Groups`, `Add Endpoint Group`, `Get Endpoint Group`, `Add Endpoint to Group`, `Remove Endpoint from Group`, `Get Allowed Items`, `Add Allowed Item`, `Remove Allowed Item`, `Get Blocked Items`, `Add Blocked Item`, `Remove Blocked Item`, `Get Endpoints in Group`
```
Comment (mgrandys-r7) [inline plugins/sophos_central/help.md:1695]:
```suggestion
* 4.5.0 - Add new action: `Get Endpoints in Group`
* 4.4.0 - Add new actions: `Isolate Endpoint`, `Get Endpoint Groups`, `Add Endpoint Group`, `Get Endpoint Group`, `Add Endpoint to Group`, `Remove Endpoint from Group`, `Get Allowed Items`, `Add Allowed Item`, `Remove Allowed Item`, `Get Blocked Items`, `Add Blocked Item`, `Remove Blocked Item`
```

Comment (mgrandys-r7) [reply plugins/sophos_central/help.md:1695]: Shouldn't we update minor?

Comment (dsliwinski-r7) [reply plugins/sophos_central/help.md:1695]: @mgrandys-r7 Version 4.4.0 has not yet been released :)

**Review by mgrandys-r7 — APPROVED** [review]

File: plugins/sophos_central/help.md [hunk 2]
```diff
@@ -58,6 +58,81 @@ Example input:
 
 ### Actions
 
+#### Get Endpoints in Group
+
+This action endpoints in your specified group.
```
Comment (dlaverty-r7) [inline plugins/sophos_central/help.md:63]: Reword: This action retrieves endpoints in your specified group.

File: plugins/sophos_central/help.md [hunk 3]
```diff
@@ -58,6 +58,81 @@ Example input:
...
+
+##### Input
+
+|Name|Type|Default|Required|Description|Enum|Example|
+|----|----|-------|--------|-----------|----|-------|
+|fields|[]string|None|False|The fields to return in a partial response. All fields are returned by default|None|["id", "type"]|
+|groupId|string|None|True|Endpoint group ID|None|9de5069c-5afe-602b-2ea0-a04b66beb2c0|
+|pageFromKey|string|None|False|The key of the item from where to fetch a page|None|exampleKey|
```
Comment (dlaverty-r7) [inline plugins/sophos_central/help.md:71]: Suggested reword: Key of page to fetch

File: plugins/sophos_central/plugin.spec.yaml [hunk 4]
```diff
@@ -1572,3 +1724,70 @@ actions:
...
+        example: ["id:asc"]
+      fields:
+        title: Fields
+        description: The fields to return in a partial response. All fields are returned by default
+        type: "[]string"
+        required: false
+        example: ["id", "type"]
+      pageFromKey:
+        title: Page from Key
+        description: The key of the item from where to fetch a page
```
Comment (dlaverty-r7) [inline plugins/sophos_central/plugin.spec.yaml:1751]: (Reword) Key of the page to fetch

**Review by llaszuk-r7 — APPROVED** [review]

**Review by dlaverty-r7 — COMMENTED** [review]
Generally lgtm, just a couple of small points to address and can you check why one of the checks are failing please?

**Review by dlaverty-r7 — APPROVED** [review]

**Comment by dlaverty-r7** [comment]
> > Generally lgtm, just a couple of small points to address and can you check why one of the checks are failing please?
> 
> @dlaverty-r7 HelpInputOutputValidator fails because it incorrectly recognizes the given timestamp example in help.md as invalid. This is a known issue with the PluginValidator.

Validator "HelpInputOutputValidator" failed! 
	Cause: plugin.spec is missing output example for {'title': 'Items', 'description': 'Endpoint items', 'type': '[]endpoint_entry', 'required': True} - is there a missing output example for this one?

**Review by dlaverty-r7 — APPROVED** [review]

**Pull request merged**